dataplane/internal/engine/engine.go

```diff
@@ -233,45 +282,249 @@ func (e *Engine) AddLayer3PuntRule(ctx context.Context, portName string, ip []by
+		nextHop := net.ParseIP(nhIPStr)
+		nextHopIP = nextHop.To4()
+		if nextHopIP == nil {
+			nextHopIP = nextHop.To16()
 		}
+	}
+	// Set the next hop IP in the packet's metadata.
+	nextHopAct := fwdbuilder.Action(fwdbuilder.UpdateAction(fwdpb.UpdateType_UPDATE_TYPE_SET, fwdpb.PacketFieldNum_PACKET_FIELD_NUM_NEXT_HOP_IP).WithValue(nextHopIP)).Build()
+	if nextHopIP == nil {
+		nextHopAct = fwdbuilder.Action(fwdbuilder.UpdateAction(fwdpb.UpdateType_UPDATE_TYPE_COPY, fwdpb.PacketFieldNum_PACKET_FIELD_NUM_NEXT_HOP_IP).WithFieldSrc(fwdpb.PacketFieldNum_PACKET_FIELD_NUM_IP_ADDR_DST)).Build()
+	}
+	// Set the output port of the packet.
+	transmitAct := fwdbuilder.Action(fwdbuilder.TransmitAction(nh.GetPort())).Build()
 
-		// If there are multiple next-hops, configure the route to use ECMP or WCMP.
-		actions = []*fwdpb.ActionDesc{{
-			ActionType: fwdpb.ActionType_ACTION_TYPE_SELECT_ACTION_LIST,
-			Action: &fwdpb.ActionDesc_Select{
-				Select: &fwdpb.SelectActionListActionDesc{
-					SelectAlgorithm: fwdpb.SelectActionListActionDesc_SELECT_ALGORITHM_CRC32, // TODO: should algo + hash be configurable?
-					FieldIds: []*fwdpb.PacketFieldId{{Field: &fwdpb.PacketField{ // Hash the traffic flow, identified, IP protocol, L3 SRC, DST address, and L4 ports (if present).
-						FieldNum: fwdpb.PacketFieldNum_PACKET_FIELD_NUM_IP_PROTO,
-					}}, {Field: &fwdpb.PacketField{
-						FieldNum: fwdpb.PacketFieldNum_PACKET_FIELD_NUM_IP_ADDR_SRC,
-					}}, {Field: &fwdpb.PacketField{
-						FieldNum: fwdpb.PacketFieldNum_PACKET_FIELD_NUM_IP_ADDR_DST,
-					}}, {Field: &fwdpb.PacketField{
-						FieldNum: fwdpb.PacketFieldNum_PACKET_FIELD_NUM_L4_PORT_SRC,
-					}}, {Field: &fwdpb.PacketField{
-						FieldNum: fwdpb.PacketFieldNum_PACKET_FIELD_NUM_L4_PORT_DST,
-					}}},
-					ActionLists: actLists,
+	acts := append([]*fwdpb.ActionDesc{nextHopAct, transmitAct}, nh.GetPreTransmitActions()...)
+	entries := &fwdpb.TableEntryAddRequest{
+		ContextId: &fwdpb.ContextId{Id: e.id},
+		TableId: &fwdpb.TableId{
+			ObjectId: &fwdpb.ObjectId{
+				Id: nhTable,
+			},
+		},
+		Entries: []*fwdpb.TableEntryAddRequest_Entry{{
+			EntryDesc: &fwdpb.EntryDesc{
+				Entry: &fwdpb.EntryDesc_Exact{
+					Exact: &fwdpb.ExactEntryDesc{
+						Fields: []*fwdpb.PacketFieldBytes{{
+							Bytes: binary.BigEndian.AppendUint64(nil, id),
+							FieldId: &fwdpb.PacketFieldId{
+								Field: &fwdpb.PacketField{
+									FieldNum: fwdpb.PacketFieldNum_PACKET_FIELD_NUM_NEXT_HOP_ID,
+								},
+							},
+						}},
+					},
 				},
 			},
-		}}
+			Actions: acts,
+		}},
+	}
+	if _, err := e.Server.TableEntryAdd(ctx, entries); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// AddIPRoute adds a route to the FIB. It operates in two modes:
+// 1. Client-managed IDs: each next hop and next hop group must be created before adding to a route with user provided ids.
+// 2. Server-managed IDs: each next hop and next hop group must be specified with route. The server implicitly creates ids.
```

Comment:
Is the mode implicit via convention or enforced in some way?

Reply:
added todo